test/middleware/express.test.ts

```diff
@@ -0,0 +1,362 @@
+            test('Adds UNIX timestamp and complexity', () => {
+                const expectedResponse = {
+                    locals: {},
+                };
+
+                middleware(mockRequest as Request, mockResponse as Response, nextFunction);
+
+                expect(mockResponse.locals).toHaveProperty('complexity');
+                expect(mockResponse.locals?.complexity).toBeInstanceOf('number');
+                expect(mockResponse.locals?.complexity).toBeGreaterThanOrEqual(0);
+
+                expect(mockResponse.locals).toHaveProperty('timestamp');
+                expect(mockResponse.locals?.timestamp).toBeInstanceOf('number');
+                // confirm that this is timestamp +/- 5 minutes of now.
+                const now: number = Date.now().valueOf();
+                const diff: number = Math.abs(now - (mockResponse.locals?.timestamp || 0));
+                expect(diff).toBeLessThan(5 * 60);
+            });
+        });
+
+        describe('Correctly limits requests', () => {
+            describe('Allows requests', () => {
+                test('...a single request', () => {
+                    // successful request calls next without any arguments.
+                    middleware(mockRequest as Request, mockResponse as Response, nextFunction);
+                    expect(nextFunction).toBeCalledTimes(1);
+                    expect(nextFunction).toBeCalledWith();
+                });
+
+                test('Multiple valid requests at > 1 second intervals', () => {
+                    for (let i = 0; i < 3; i++) {
+                        const next: NextFunction = jest.fn();
+                        middleware(complexRequest as Request, mockResponse as Response, next);
+                        expect(next).toBeCalledTimes(1);
+                        expect(next).toBeCalledWith();
+
+                        // advance the timers by 1 second for the next request
+                        jest.advanceTimersByTime(1000);
+                    }
+                });
+
+                test('Multiple valid requests at within one second', () => {
+                    for (let i = 0; i < 3; i++) {
+                        const next: NextFunction = jest.fn();
+                        middleware(complexRequest as Request, mockResponse as Response, next);
+                        expect(next).toBeCalledTimes(1);
+                        expect(next).toBeCalledWith();
+
+                        // advance the timers by 1 second for the next request
+                        jest.advanceTimersByTime(20);
+                    }
+                });
+            });
+
+            describe('BLOCKS requests', () => {
+                test('A single request that exceeds capacity', () => {
+                    complexRequest = {
+                        // complexity should be 12 if 'first' is accounted for.
+                        // scalars: 1, droid: 1, reviews (5 * (1 Review, 1 episode))
+                        query: {
```

Comment:
I'm reading a complexity of 8 with 1 `query`, 1 `scalars`, 1 `droid`, and 5 `reviews`. Episode is an enum is has weight 0

Reply:
This needs to be greater than 10 for the tests to work.